expression/builtin_control.go

```diff
@@ -61,47 +62,69 @@ func maxlen(lhsFlen, rhsFlen int) int {
-		// If both arguments are NULL, make resulting type BINARY(0).
-		if rhs.GetType() == mysql.TypeNull {
-			resultFieldType.SetType(mysql.TypeString)
-			resultFieldType.SetFlen(0)
-			resultFieldType.SetDecimal(0)
-			types.SetBinChsClnFlag(resultFieldType)
+func setFlenFromArgs(evalType types.EvalType, resultFieldType *types.FieldType, argTps ...*types.FieldType) {
+	if evalType == types.ETDecimal || evalType == types.ETInt {
+		maxArgFlen := 0
+		for i := range argTps {
+			flagLen := 0
+			if !mysql.HasUnsignedFlag(argTps[i].GetFlag()) {
+				flagLen = 1
+			}
+			flen := argTps[i].GetFlen() - flagLen
+			if argTps[i].GetDecimal() != types.UnspecifiedLength {
+				flen -= argTps[i].GetDecimal()
+			}
+			maxArgFlen = maxlen(maxArgFlen, flen)
 		}
-	} else if rhs.GetType() == mysql.TypeNull {
-		*resultFieldType = *lhs
-		tempFlag := resultFieldType.GetFlag()
-		types.SetTypeFlag(&tempFlag, mysql.NotNullFlag, false)
-		resultFieldType.SetFlag(tempFlag)
+		resultFlen := maxArgFlen + resultFieldType.GetDecimal() + 1 // account for -1 len fields
+		resultFieldType.SetFlenUnderLimit(resultFlen)
+	} else if evalType == types.ETString {
+		maxLen := 0
+		for i := range argTps {
+			argFlen := argTps[i].GetFlen()
+			if argFlen == types.UnspecifiedLength {
+				return
+			}
+			maxLen = maxlen(argFlen, maxLen)
+		}
+		resultFieldType.SetFlen(maxLen)
 	} else {
-		resultFieldType = types.AggFieldType([]*types.FieldType{lhs, rhs})
-		var tempFlag uint
-		evalType := types.AggregateEvalType([]*types.FieldType{lhs, rhs}, &tempFlag)
-		resultFieldType.SetFlag(tempFlag)
-		if evalType == types.ETInt {
-			resultFieldType.SetDecimal(0)
-		} else {
-			if lhs.GetDecimal() == types.UnspecifiedLength || rhs.GetDecimal() == types.UnspecifiedLength {
+		maxLen := 0
+		for i := range argTps {
+			maxLen = maxlen(argTps[i].GetFlen(), maxLen)
+		}
+		resultFieldType.SetFlen(maxLen)
+	}
+}
+
+func setDecimalFromArgs(evalType types.EvalType, resultFieldType *types.FieldType, argTps ...*types.FieldType) {
+	if evalType == types.ETInt {
+		resultFieldType.SetDecimal(0)
+	} else {
+		maxDecimal := 0
+		for i := range argTps {
+			if argTps[i].GetDecimal() == types.UnspecifiedLength {
 				resultFieldType.SetDecimal(types.UnspecifiedLength)
-			} else {
-				resultFieldType.SetDecimalUnderLimit(mathutil.Max(lhs.GetDecimal(), rhs.GetDecimal()))
+				return
 			}
+			maxDecimal = mathutil.Max(argTps[i].GetDecimal(), maxDecimal)
 		}
+		resultFieldType.SetDecimalUnderLimit(maxDecimal)
+	}
+}
 
+func setCollateAndCharsetAndFlagFromArgs(ctx sessionctx.Context, funcName string, evalType types.EvalType, resultFieldType *types.FieldType, args ...Expression) error {
+	switch funcName {
+	case ast.If, ast.Ifnull, ast.WindowFuncLead, ast.WindowFuncLag:
+		if len(args) != 2 {
+			panic("unexpected length of args for if/ifnull/lead/lag")
+		}
+		lexp, rexp := args[0], args[1]
+		lhs, rhs := lexp.GetType(), rexp.GetType()
 		if types.IsNonBinaryStr(lhs) && !types.IsBinaryStr(rhs) {
```

Comment:
how about merge the branch of `if types.IsNonBinaryStr(lhs) && !types.IsBinaryStr(rhs) {} else if types.IsNonBinaryStr(rhs) && !types.IsBinaryStr(lhs) {}` into the following
```
if types.IsNonBinaryStr(lhs) || types.IsNonBinaryStr(rhs) {
	ec, err := CheckAndDeriveCollationFromExprs(ctx, funcName, evalType, lexp, rexp)
	if err != nil {
		return nil, err
	}
	resultFieldType.SetCollate(ec.Collation)
	resultFieldType.SetCharset(ec.Charset)
	resultFieldType.SetFlag(0)
	if (!types.IsNonBinaryStr(lhs) || mysql.HasBinaryFlag(lhs.GetFlag())) || (!types.IsNonBinaryStr(rhs) && mysql.HasBinaryFlag(rhs.GetFlag())) {
		resultFieldType.AddFlag(mysql.BinaryFlag)
	}
} 
```

Reply:
It seems that the logic of the two is different...

Reply:
ok, refined.